Groovy: fails to parse try-with-resources without 'catch' block

<!--
Thank you for reporting an issue with OpenRewrite!
We appreciate you taking the time to help us improve.
Please fill out the template below to help us understand and reproduce the issue.
Feel free to delete any sections that don't apply to your issue.
-->

## What version of OpenRewrite are you using?
- Running test from main branch https://github.com/gtiwari333/rewrite/pull/2/changes
 
## How are you running OpenRewrite?
- Running the new tests from https://github.com/gtiwari333/rewrite/pull/2/changes

## What is the smallest, simplest way to reproduce the problem?
 - https://github.com/gtiwari333/rewrite/pull/2/changes

Run these tests from `org.openrewrite.groovy.tree.TryTest`

```groovy
    @Test
    void tryWithResourcesWithoutCatch() {
        rewriteRun(
          groovy(
            """
              try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              
              }
              """
          )
        );
    }

    @Test
    void tryWithResourcesWithoutCatchNested() {
        rewriteRun(
          groovy(
            """
              try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
                            try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream d = new ByteArrayInputStream("".getBytes())) {
              
                             }
              }
              """
          )
        );
    }

    @Test
    void tryWithResourcesWithoutCatchNestedMultiline() {
        rewriteRun(
          groovy(
            """
              try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes());
                  ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
                            try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes());
                                ByteArrayInputStream d = new ByteArrayInputStream("".getBytes())) {
              
                             }
              }
              """
          )
        );
    }
 
```
## What did you expect to see?
No failures during parsing the code that has try-with-resources 

## What did you see instead?
<!-- A code snippet, or a description of the behavior you saw instead of the above expected result. -->
 Error similar to following

```
Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes());
    ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes());
                  ByteAr
org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes());
    ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes());
                  ByteAr
```

## What is the full stack trace of any errors you encountered?
<!-- When errors occur, please include the output of `--stacktrace` for Gradle or `--debug` for Maven. -->
```

Failed to parse sources or run recipe
java.lang.AssertionError: Failed to parse sources or run recipe
	at org.openrewrite.test.RewriteTest.lambda$defaultExecutionContext$14(RewriteTest.java:722)
	at org.openrewrite.test.RewriteTest$$Lambda/0x00007d54741b4660.accept(Unknown Source)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:186)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatchNestedMultiline(TryTest.java:182)
Caused by: org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes());
    ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes());
                  ByteAr
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:248)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	... 13 more
Caused by: java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@4665428b[org.codehaus.groovy.ast.stmt.ExpressionStatement@55e1192[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@1213ffbc[org.codehaus.groovy.ast.expr.VariableExpression@16b3c905[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@1ef31f71[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6e8fdd19[org.codehaus.groovy.ast.expr.MethodCallExpression@35787726[object: org.codehaus.groovy.ast.expr.ConstantExpression@2199e845[] method: org.codehaus.groovy.ast.expr.ConstantExpression@27976390[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@37e0056e[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@415795f3[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2fe2965c[org.codehaus.groovy.ast.expr.VariableExpression@49fdbe2b[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@3375ebd3[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@40943a6]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	... 14 more


Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes());
    ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes());
                  ByteAr
org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes());
    ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes());
                  ByteAr
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:248)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatchNestedMultiline(TryTest.java:182)
Caused by: java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@4665428b[org.codehaus.groovy.ast.stmt.ExpressionStatement@55e1192[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@1213ffbc[org.codehaus.groovy.ast.expr.VariableExpression@16b3c905[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@1ef31f71[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6e8fdd19[org.codehaus.groovy.ast.expr.MethodCallExpression@35787726[object: org.codehaus.groovy.ast.expr.ConstantExpression@2199e845[] method: org.codehaus.groovy.ast.expr.ConstantExpression@27976390[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@37e0056e[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@415795f3[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2fe2965c[org.codehaus.groovy.ast.expr.VariableExpression@49fdbe2b[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@3375ebd3[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@40943a6]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	... 14 more


Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@4665428b[org.codehaus.groovy.ast.stmt.ExpressionStatement@55e1192[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@1213ffbc[org.codehaus.groovy.ast.expr.VariableExpression@16b3c905[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@1ef31f71[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6e8fdd19[org.codehaus.groovy.ast.expr.MethodCallExpression@35787726[object: org.codehaus.groovy.ast.expr.ConstantExpression@2199e845[] method: org.codehaus.groovy.ast.expr.ConstantExpression@27976390[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@37e0056e[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@415795f3[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2fe2965c[org.codehaus.groovy.ast.expr.VariableExpression@49fdbe2b[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@3375ebd3[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@40943a6]}
java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@4665428b[org.codehaus.groovy.ast.stmt.ExpressionStatement@55e1192[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@1213ffbc[org.codehaus.groovy.ast.expr.VariableExpression@16b3c905[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@1ef31f71[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6e8fdd19[org.codehaus.groovy.ast.expr.MethodCallExpression@35787726[object: org.codehaus.groovy.ast.expr.ConstantExpression@2199e845[] method: org.codehaus.groovy.ast.expr.ConstantExpression@27976390[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@37e0056e[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@415795f3[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2fe2965c[org.codehaus.groovy.ast.expr.VariableExpression@49fdbe2b[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@3375ebd3[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@40943a6]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatchNestedMultiline(TryTest.java:182)


Failed to parse sources or run recipe
java.lang.AssertionError: Failed to parse sources or run recipe
	at org.openrewrite.test.RewriteTest.lambda$defaultExecutionContext$14(RewriteTest.java:722)
	at org.openrewrite.test.RewriteTest$$Lambda/0x00007d54741b4660.accept(Unknown Source)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:186)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatch(TryTest.java:154)
Caused by: org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {

}
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:248)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	... 13 more
Caused by: java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@7a64cb0c[org.codehaus.groovy.ast.stmt.ExpressionStatement@6c4ce583[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@755b5f30[org.codehaus.groovy.ast.expr.VariableExpression@6f1fa1d0[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@29bbc63c[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@272778ae[org.codehaus.groovy.ast.expr.MethodCallExpression@2cccf134[object: org.codehaus.groovy.ast.expr.ConstantExpression@3c18942[] method: org.codehaus.groovy.ast.expr.ConstantExpression@743c3520[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6842c101[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@234cd86c[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2c48cede[org.codehaus.groovy.ast.expr.VariableExpression@28b523a[variable: __$$primaryExc0 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@39c87b42[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@47fcefb3]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	... 14 more


Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {

}
org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {

}
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:248)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatch(TryTest.java:154)
Caused by: java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@7a64cb0c[org.codehaus.groovy.ast.stmt.ExpressionStatement@6c4ce583[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@755b5f30[org.codehaus.groovy.ast.expr.VariableExpression@6f1fa1d0[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@29bbc63c[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@272778ae[org.codehaus.groovy.ast.expr.MethodCallExpression@2cccf134[object: org.codehaus.groovy.ast.expr.ConstantExpression@3c18942[] method: org.codehaus.groovy.ast.expr.ConstantExpression@743c3520[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6842c101[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@234cd86c[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2c48cede[org.codehaus.groovy.ast.expr.VariableExpression@28b523a[variable: __$$primaryExc0 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@39c87b42[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@47fcefb3]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	... 14 more


Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@7a64cb0c[org.codehaus.groovy.ast.stmt.ExpressionStatement@6c4ce583[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@755b5f30[org.codehaus.groovy.ast.expr.VariableExpression@6f1fa1d0[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@29bbc63c[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@272778ae[org.codehaus.groovy.ast.expr.MethodCallExpression@2cccf134[object: org.codehaus.groovy.ast.expr.ConstantExpression@3c18942[] method: org.codehaus.groovy.ast.expr.ConstantExpression@743c3520[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6842c101[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@234cd86c[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2c48cede[org.codehaus.groovy.ast.expr.VariableExpression@28b523a[variable: __$$primaryExc0 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@39c87b42[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@47fcefb3]}
java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@7a64cb0c[org.codehaus.groovy.ast.stmt.ExpressionStatement@6c4ce583[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@755b5f30[org.codehaus.groovy.ast.expr.VariableExpression@6f1fa1d0[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@29bbc63c[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@272778ae[org.codehaus.groovy.ast.expr.MethodCallExpression@2cccf134[object: org.codehaus.groovy.ast.expr.ConstantExpression@3c18942[] method: org.codehaus.groovy.ast.expr.ConstantExpression@743c3520[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@6842c101[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@234cd86c[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@2c48cede[org.codehaus.groovy.ast.expr.VariableExpression@28b523a[variable: __$$primaryExc0 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@39c87b42[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@47fcefb3]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatch(TryTest.java:154)


Failed to parse sources or run recipe
java.lang.AssertionError: Failed to parse sources or run recipe
	at org.openrewrite.test.RewriteTest.lambda$defaultExecutionContext$14(RewriteTest.java:722)
	at org.openrewrite.test.RewriteTest$$Lambda/0x00007d54741b4660.accept(Unknown Source)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:186)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatchNested(TryTest.java:167)
Caused by: org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream d = new
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:248)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	... 13 more
Caused by: java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@2f15a964[org.codehaus.groovy.ast.stmt.ExpressionStatement@2d70f312[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@4c7d7430[org.codehaus.groovy.ast.expr.VariableExpression@7d25913[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@4ea1aa52[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@74123110[org.codehaus.groovy.ast.expr.MethodCallExpression@5ed0b4e3[object: org.codehaus.groovy.ast.expr.ConstantExpression@79ab97fd[] method: org.codehaus.groovy.ast.expr.ConstantExpression@3a9b41a0[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@5fe46d52[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@448ade1[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@21f9c6ea[org.codehaus.groovy.ast.expr.VariableExpression@c14e5a5[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@418f0534[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@3dfd7eaa]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	... 14 more


Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream d = new
org.openrewrite.groovy.GroovyParsingException: Failed to parse file.groovy at cursor position 0. The surrounding characters in the original source are:
~cursor~>try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
              try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream d = new
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:248)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatchNested(TryTest.java:167)
Caused by: java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@2f15a964[org.codehaus.groovy.ast.stmt.ExpressionStatement@2d70f312[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@4c7d7430[org.codehaus.groovy.ast.expr.VariableExpression@7d25913[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@4ea1aa52[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@74123110[org.codehaus.groovy.ast.expr.MethodCallExpression@5ed0b4e3[object: org.codehaus.groovy.ast.expr.ConstantExpression@79ab97fd[] method: org.codehaus.groovy.ast.expr.ConstantExpression@3a9b41a0[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@5fe46d52[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@448ade1[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@21f9c6ea[org.codehaus.groovy.ast.expr.VariableExpression@c14e5a5[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@418f0534[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@3dfd7eaa]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	... 14 more


Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@2f15a964[org.codehaus.groovy.ast.stmt.ExpressionStatement@2d70f312[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@4c7d7430[org.codehaus.groovy.ast.expr.VariableExpression@7d25913[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@4ea1aa52[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@74123110[org.codehaus.groovy.ast.expr.MethodCallExpression@5ed0b4e3[object: org.codehaus.groovy.ast.expr.ConstantExpression@79ab97fd[] method: org.codehaus.groovy.ast.expr.ConstantExpression@3a9b41a0[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@5fe46d52[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@448ade1[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@21f9c6ea[org.codehaus.groovy.ast.expr.VariableExpression@c14e5a5[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@418f0534[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@3dfd7eaa]}
java.lang.IllegalStateException: Expected to find a parent for Cursor{org.codehaus.groovy.ast.stmt.BlockStatement@2f15a964[org.codehaus.groovy.ast.stmt.ExpressionStatement@2d70f312[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@4c7d7430[org.codehaus.groovy.ast.expr.VariableExpression@7d25913[variable: a type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream]("=" at 1:27:  "=" )org.codehaus.groovy.ast.expr.ConstructorCallExpression@4ea1aa52[type: java.io.ByteArrayInputStream -> java.io.ByteArrayInputStream arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@74123110[org.codehaus.groovy.ast.expr.MethodCallExpression@5ed0b4e3[object: org.codehaus.groovy.ast.expr.ConstantExpression@79ab97fd[] method: org.codehaus.groovy.ast.expr.ConstantExpression@3a9b41a0[getBytes] arguments: org.codehaus.groovy.ast.expr.ArgumentListExpression@5fe46d52[]]]]]], org.codehaus.groovy.ast.stmt.ExpressionStatement@448ade1[expression:org.codehaus.groovy.ast.expr.DeclarationExpression@21f9c6ea[org.codehaus.groovy.ast.expr.VariableExpression@c14e5a5[variable: __$$primaryExc2 type: java.lang.Throwable -> java.lang.Throwable]("=":  "=" )org.codehaus.groovy.ast.expr.ConstantExpression@418f0534[null]]], org.codehaus.groovy.ast.stmt.TryCatchStatement@3dfd7eaa]}
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:235)
	at org.openrewrite.Cursor.getParentOrThrow(Cursor.java:241)
	at org.openrewrite.groovy.GroovyParserVisitor$RewriteGroovyVisitor.visitBlockStatement(GroovyParserVisitor.java:1749)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.openrewrite.groovy.GroovyParserVisitor.convertTopLevelStatement(GroovyParserVisitor.java:3476)
	at org.openrewrite.groovy.GroovyParserVisitor.visit(GroovyParserVisitor.java:234)
	at org.openrewrite.groovy.GroovyParser.lambda$parseInputs$3(GroovyParser.java:175)
	at org.openrewrite.groovy.GroovyParser$$Lambda/0x00007d5474201798.apply(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:325)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:159)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:154)
	at org.openrewrite.groovy.tree.TryTest.tryWithResourcesWithoutCatchNested(TryTest.java:167)



```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
<!-- Indicate if this is something you would like to work on, and how we can best support you in doing so. -->

I can with LLM agents help.

--- a/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
+++ b/rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java
@@ -1744,6 +1744,15 @@ public class GroovyParserVisitor {
 
         @Override
         public void visitBlockStatement(BlockStatement block) {
+            if (isUnwrappedDesugaredResourceBlock(block)) {
+                // A try-with-resources without a catch or finally clause is desugared by Groovy 4 into a
+                // bare block (rather than being wrapped in a TryCatchStatement like the cases that have a
+                // catch/finally). Re-wrap it so the try-with-resources handling in visitTryCatchFinally applies.
+                BlockStatement wrapper = new BlockStatement();
+                wrapper.addStatement(block);
+                visitTryCatchFinally(new TryCatchStatement(wrapper, EmptyStatement.INSTANCE));
+                return;
+            }
             Space fmt = EMPTY;
             Space staticInitPadding = EMPTY;
             boolean isStaticInit = sourceStartsWith("static");
@@ -3160,6 +3169,9 @@ public class GroovyParserVisitor {
             // Groovy 4 desugars try-with-resources at parse time (getResourceStatements() is always empty).
             // Detect from source: if "(" follows "try", parse resources from source text.
             JContainer<J.Try.Resource> resources = null;
+            // The body lives at the bottom of the nested resource structure; it defaults to the try
+            // statement itself when there are no resources, and is reassigned as resources are consumed.
+            org.codehaus.groovy.ast.stmt.Statement bodyStatement = node.getTryStatement();
             boolean hasTryWithResources = source.charAt(indexOfNextNonWhitespace(cursor, source)) == '(';
             if (hasTryWithResources) {
                 Space beforeParen = sourceBefore("(");
@@ -3177,7 +3189,6 @@ public class GroovyParserVisitor {
                     resourceVar = resourceVar.withPrefix(EMPTY);
 
                     TryCatchStatement innerTry = (TryCatchStatement) innerBlock.getStatements().get(innerBlock.getStatements().size() - 1);
-                    boolean hasMoreResources = isDesugaredResourceBlock(innerTry.getTryStatement());
 
                     int nextNonWs = indexOfNextNonWhitespace(cursor, source);
                     boolean semicolonPresent = nextNonWs < source.length() && source.charAt(nextNonWs) == ';';
@@ -3191,36 +3202,23 @@ public class GroovyParserVisitor {
                             resourceVar.withPrefix(EMPTY), semicolonPresent);
                     skip(";");
 
+                    // Whether another resource follows must be decided from the source, not the AST:
+                    // a nested try-with-resources in the body desugars to the same shape as an additional
+                    // resource, so only the closing ')' of the resource list can disambiguate them.
+                    boolean hasMoreResources = source.charAt(indexOfNextNonWhitespace(cursor, source)) != ')';
                     if (hasMoreResources) {
                         resourceList.add(padRight(tryResource, EMPTY));
                         current = innerTry.getTryStatement();
                     } else {
                         resourceList.add(padRight(tryResource, sourceBefore(")")));
+                        bodyStatement = innerTry.getTryStatement();
                         break;
                     }
                 }
                 resources = JContainer.build(beforeParen, resourceList, Markers.EMPTY);
             }
 
-            // When try-with-resources, find the actual body by walking down the nested structure
-            // to the innermost TryCatchStatement's tryStmt.
-            J.Block body;
-            if (hasTryWithResources) {
-                org.codehaus.groovy.ast.stmt.Statement current = node.getTryStatement();
-                TryCatchStatement innerTry = null;
-                while (isDesugaredResourceBlock(current)) {
-                    BlockStatement innerBlock = (BlockStatement) ((BlockStatement) current).getStatements().get(0);
-                    innerTry = (TryCatchStatement) innerBlock.getStatements().get(innerBlock.getStatements().size() - 1);
-                    if (isDesugaredResourceBlock(innerTry.getTryStatement())) {
-                        current = innerTry.getTryStatement();
-                    } else {
-                        break;
-                    }
-                }
-                body = doVisit(innerTry != null ? innerTry.getTryStatement() : node.getTryStatement());
-            } else {
-                body = doVisit(node.getTryStatement());
-            }
+            J.Block body = doVisit(bodyStatement);
 
             // Handle catches, merging multi-catch statements.
             // Groovy 4 splits catch(A | B e) into separate CatchStatements at the same source position.
@@ -3631,6 +3629,27 @@ public class GroovyParserVisitor {
 
     private <T> JRightPadded<T> padRight(T tree, Space right, Markers markers) {
         return new JRightPadded<>(tree, right, markers);
+    }
+
+    /**
+     * Detects the "unwrapped" desugared form of a try-with-resources that has no catch or finally clause.
+     * Groovy 4 desugars such a statement into a bare block of the shape
+     * {@code [resourceDecl, __$$primaryExc decl, TryCatchStatement]} rather than wrapping it in an
+     * enclosing {@link TryCatchStatement}. The synthetic {@code __$$primaryExc} primary-exception holder is
+     * what distinguishes it from an ordinary user-authored block.
+     */
+    private static boolean isUnwrappedDesugaredResourceBlock(BlockStatement block) {
+        List<org.codehaus.groovy.ast.stmt.Statement> stmts = block.getStatements();
+        if (stmts.size() < 3 || !(stmts.get(stmts.size() - 1) instanceof TryCatchStatement)) {
+            return false;
+        }
+        org.codehaus.groovy.ast.stmt.Statement primaryExc = stmts.get(stmts.size() - 2);
+        if (primaryExc instanceof ExpressionStatement &&
+                ((ExpressionStatement) primaryExc).getExpression() instanceof DeclarationExpression) {
+            org.codehaus.groovy.ast.expr.Expression left = ((DeclarationExpression) ((ExpressionStatement) primaryExc).getExpression()).getLeftExpression();
+            return left instanceof VariableExpression && ((VariableExpression) left).getName().startsWith("__$$primaryExc");
+        }
+        return false;
     }
 
     /**

--- a/rewrite-groovy/src/test/java/org/openrewrite/groovy/tree/TryTest.java
+++ b/rewrite-groovy/src/test/java/org/openrewrite/groovy/tree/TryTest.java
@@ -167,6 +167,65 @@ class TryTest implements RewriteTest {
         );
     }
 
+    @Issue("https://github.com/openrewrite/rewrite/issues/8040")
+    @Test
+    void tryWithResourcesWithoutCatch() {
+        rewriteRun(
+          groovy(
+            """
+              try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
+
+              }
+              """
+          )
+        );
+    }
+
+    @Issue("https://github.com/openrewrite/rewrite/issues/8040")
+    @Test
+    void tryWithResourcesWithoutCatchNested() {
+        rewriteRun(
+          groovy(
+            """
+              try(ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
+                            try(ByteArrayInputStream c = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream d = new ByteArrayInputStream("".getBytes())) {
+
+                             }
+              }
+              """
+          )
+        );
+    }
+
+    @Issue("https://github.com/openrewrite/rewrite/issues/8040")
+    @Test
+    void tryWithResourceWithoutCatch() {
+        rewriteRun(
+          groovy(
+            """
+              try (ByteArrayInputStream a = new ByteArrayInputStream("".getBytes())) {
+                  def x = a.read()
+              }
+              """
+          )
+        );
+    }
+
+    @Issue("https://github.com/openrewrite/rewrite/issues/8040")
+    @Test
+    void tryWithResourcesWithoutCatchButFinally() {
+        rewriteRun(
+          groovy(
+            """
+              try (ByteArrayInputStream a = new ByteArrayInputStream("".getBytes()); ByteArrayInputStream b = new ByteArrayInputStream("".getBytes())) {
+              } finally {
+                  println "done"
+              }
+              """
+          )
+        );
+    }
+
     @Test
     void whitespace() {
         rewriteRun(

